File: tools/doubao_visual_lesson_placement.mjs
/**
 * Purpose:
 * - Calls the Doubao model to place visual lesson material into candidate curriculum buckets.
 * - Keep request shaping and result packaging together here so prompt changes are auditable.
 */

import fs from "node:fs/promises";
import path from "node:path";

const DEFAULT_MODEL = "doubao-seed-2-0-pro-260215";
const API_URL = "https://ark.cn-beijing.volces.com/api/v3/chat/completions";

function parseArgs(argv) {
  const args = {};
  for (let i = 0; i < argv.length; i += 1) {
    const key = argv[i];
    if (!key.startsWith("--")) continue;
    const next = argv[i + 1];
    if (!next || next.startsWith("--")) {
      args[key.slice(2)] = true;
    } else {
      args[key.slice(2)] = next;
      i += 1;
    }
  }
  return args;
}

function normalizeText(text) {
  return String(text || "").replace(/\s+/g, " ").trim();
}

function extractJsonObject(text) {
  const raw = String(text || "").trim().replace(/^```(?:json)?\s*/i, "").replace(/\s*```$/, "");
  try {
    return JSON.parse(raw);
  } catch {}

  const start = raw.indexOf("{");
  if (start < 0) throw new Error("json_object_not_found");
  let depth = 0;
  let inString = false;
  let escape = false;
  for (let i = start; i < raw.length; i += 1) {
    const ch = raw[i];
    if (inString) {
      if (escape) escape = false;
      else if (ch === "\\") escape = true;
      else if (ch === '"') inString = false;
      continue;
    }
    if (ch === '"') inString = true;
    else if (ch === "{") depth += 1;
    else if (ch === "}") {
      depth -= 1;
      if (depth === 0) return JSON.parse(raw.slice(start, i + 1));
    }
  }
  throw new Error("json_object_not_closed");
}

async function readJson(filePath) {
  return JSON.parse(await fs.readFile(filePath, "utf8"));
}

async function ensureDir(dirPath) {
  await fs.mkdir(dirPath, { recursive: true });
}

function buildCandidateLine(point) {
  return `${point.knowledge_id}|${point.level_2_module}|${point.level_3_min_knowledge_point}`;
}

function buildPrompt(lesson, knowledgePoints, question) {
  const candidateLines = knowledgePoints.map(buildCandidateLine).join("；");
  return [
    `题目来自${lesson.stage}数学/${lesson.grade}/${lesson.season}/${lesson.lesson_title}。`,
    `候选知识点：${candidateLines}。`,
    `题目OCR：${normalizeText(question.text_preview || "")}。`,
    `讲义标签：${question.checkpoint || ""}；组件：${question.component_label || ""}；题号：${question.local_number || ""}。`,
    "请给出层级判断。",
    '只输出json：{"stage":"初中|高中","stage_reason":"<=40字","grade":"...","grade_reason":"<=40字","lesson_id":"...","lesson_reason":"<=40字","top3":["knowledge_id1","knowledge_id2","knowledge_id3"],"confidence":"high|medium|low","review_status":"accepted_candidate|needs_teacher_review","reason":"<=80字"}',
  ].join("");
}

async function callDoubao({ apiKey, model, prompt }) {
  const payload = {
    model,
    temperature: 0.1,
    messages: [
      { role: "system", content: "你是数学教研助手，只输出JSON。" },
      { role: "user", content: prompt },
    ],
  };

  const response = await fetch(API_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify(payload),
  });

  const text = await response.text();
  if (!response.ok) {
    throw new Error(`http_${response.status}: ${text}`);
  }
  const data = JSON.parse(text);
  const content = data?.choices?.[0]?.message?.content || "";
  return { raw: data, parsed: extractJsonObject(content) };
}

async function callDoubaoWithRetry({ apiKey, model, prompt, retries = 3, backoffMs = 1600 }) {
  let lastError = null;
  for (let attempt = 1; attempt <= retries; attempt += 1) {
    try {
      return await callDoubao({ apiKey, model, prompt });
    } catch (error) {
      lastError = error;
      if (attempt >= retries) break;
      await new Promise((resolve) => setTimeout(resolve, backoffMs * attempt));
    }
  }
  throw lastError || new Error("unknown_fetch_failure");
}

function toTop3Records(parsed, pointMap) {
  const top3 = Array.isArray(parsed.top3) ? parsed.top3.slice(0, 3) : [];
  return top3.map((knowledgeId, index) => {
    const point = pointMap.get(knowledgeId) || {};
    return {
      rank: index + 1,
      knowledge_id: knowledgeId,
      grade: parsed.grade || point.grade || "",
      lesson_title: point.lesson_title || "",
      module: point.level_2_module || "",
      min_knowledge_point: point.level_3_min_knowledge_point || "",
      confidence: index === 0 ? parsed.confidence || "medium" : "medium",
      teacher_review_note:
        index === 0
          ? parsed.reason || "模型给出首选落位。"
          : `作为备选保留，便于教师复核同课次内的相近知识点。`,
    };
  });
}

function buildStructuredPlacement(parsed, lesson, question, pointMap) {
  const top3Records = toTop3Records(parsed, pointMap);
  const stagePrediction = parsed.stage || lesson.stage;
  const gradePrediction = parsed.grade || lesson.grade;
  const lessonPrediction = parsed.lesson_id || lesson.lesson_id;
  const fineCandidates = top3Records.map((item, index) => ({
    knowledge_id: item.knowledge_id,
    module: item.module,
    min_knowledge_point: item.min_knowledge_point,
    confidence: index === 0 ? parsed.confidence || "medium" : "medium",
    reason:
      index === 0
        ? parsed.reason || "模型将其作为首选最小知识点。"
        : `保留为同课次内的相近备选知识点。`,
  }));

  const topModule = fineCandidates[0]?.module || "";

  return {
    question_id: question.question_id,
    question_reading: {
      core_objects: [question.checkpoint || "题块", question.component_label || "题目"],
      conditions: [
        `讲义标签：${question.checkpoint || "未标注"}`,
        `组件：${question.component_label || "未标注"}`,
        `OCR摘要：${normalizeText(question.text_preview || "").slice(0, 140)}`,
      ],
      asked_result: "将题目落到当前课次的最小知识点",
      visual_dependency: "formula",
    },
    layered_trace: {
      stage: {
        prediction: stagePrediction,
        confidence: parsed.confidence || "medium",
        reason: parsed.stage_reason || `模型判定该题属于${stagePrediction}数学范畴。`,
      },
      grade: {
        top_candidates: [
          {
            grade: gradePrediction,
            confidence: parsed.confidence || "medium",
            reason: parsed.grade_reason || `模型判定该题更贴合${gradePrediction}内容边界。`,
          },
        ],
      },
      lesson: {
        top_candidates: [
          {
            lesson_id: lessonPrediction,
            lesson_title: lesson.lesson_title,
            confidence: parsed.confidence || "medium",
            reason: parsed.lesson_reason || `模型判定该题落在《${lesson.lesson_title}》课次内。`,
          },
        ],
      },
      module: {
        top_candidates: topModule
          ? [
              {
                module: topModule,
                confidence: parsed.confidence || "medium",
                reason: `首选知识点位于模块「${topModule}」。`,
              },
            ]
          : [],
      },
      fine_point: {
        top_candidates: fineCandidates,
      },
    },
    final_top3: top3Records,
    review_status: parsed.review_status || "needs_teacher_review",
  };
}

function buildCompactRecord(question, structured) {
  const top1 = structured.final_top3?.[0] || {};
  return {
    question_id: question.question_id,
    checkpoint: question.checkpoint || "",
    status: "ok",
    top1_knowledge_id: top1.knowledge_id || "",
    top1_module: top1.module || "",
    top1_min_knowledge_point: top1.min_knowledge_point || "",
    top1_confidence: top1.confidence || "",
    review_status: structured.review_status || "",
  };
}

async function writeSummary(outPath, lesson, knowledgePoints, records) {
  const pointMap = new Map(knowledgePoints.map((point) => [point.knowledge_id, point]));
  const counter = new Map();
  const needsReview = [];

  for (const record of records) {
    const top1 = record.placement?.final_top3?.[0];
    if (top1?.knowledge_id) {
      counter.set(top1.knowledge_id, (counter.get(top1.knowledge_id) || 0) + 1);
    }
    if (record.placement?.review_status !== "accepted_candidate") {
      needsReview.push(record);
    }
  }

  const lines = [
    "# 豆包落位结果\n\n",
    `- 课次：${lesson.lesson_title}（${lesson.lesson_id}）\n`,
    `- 题量：${records.length}\n`,
    `- 需要教师复核：${needsReview.length}\n\n`,
    "## Top1 分布\n\n",
  ];

  [...counter.entries()]
    .sort((a, b) => b[1] - a[1])
    .forEach(([knowledgeId, count]) => {
      const point = pointMap.get(knowledgeId) || {};
      lines.push(`- ${knowledgeId} | ${point.level_2_module || ""} / ${point.level_3_min_knowledge_point || ""} ：${count}题\n`);
    });

  lines.push("\n## 复核清单\n\n");
  if (!needsReview.length) {
    lines.push("- 本轮无强制复核题。\n");
  } else {
    for (const record of needsReview) {
      const top1 = record.placement?.final_top3?.[0] || {};
      lines.push(`- ${record.question_id} | ${record.checkpoint} | ${top1.knowledge_id || ""} | ${top1.teacher_review_note || ""}\n`);
    }
  }

  await fs.writeFile(outPath, lines.join(""), "utf8");
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const apiKey = args["api-key"] || process.env.ARK_API_KEY || "";
  if (!apiKey) throw new Error("missing_api_key");

  const splitJson = args["split-json"];
  const knowledgeJson = args["knowledge-json"];
  const lessonsJson = args["lessons-json"];
  const lessonId = args["lesson-id"];
  const outDir = args["out-dir"];
  const model = args.model || DEFAULT_MODEL;
  const limit = Number(args.limit || 0);
  const sleepMs = Number(args["sleep-ms"] || 250);
  const retries = Number(args.retries || 3);
  const onlyQuestionIds = String(args["question-ids"] || "")
    .split(",")
    .map((item) => item.trim())
    .filter(Boolean);

  if (!splitJson || !knowledgeJson || !lessonsJson || !lessonId || !outDir) {
    throw new Error("missing_required_args");
  }

  await ensureDir(outDir);
  const rawDir = path.join(outDir, "raw");
  await ensureDir(rawDir);

  const splitData = await readJson(splitJson);
  const allPoints = await readJson(knowledgeJson);
  const allLessons = await readJson(lessonsJson);
  const lesson = allLessons.find((item) => item.lesson_id === lessonId);
  if (!lesson) throw new Error(`lesson_not_found:${lessonId}`);
  const knowledgePoints = allPoints.filter((item) => item.lesson_id === lessonId);
  if (!knowledgePoints.length) throw new Error(`knowledge_points_not_found:${lessonId}`);
  const pointMap = new Map(knowledgePoints.map((point) => [point.knowledge_id, point]));

  let questions = limit > 0 ? splitData.questions.slice(0, limit) : splitData.questions;
  if (onlyQuestionIds.length) {
    const questionSet = new Set(onlyQuestionIds);
    questions = questions.filter((item) => questionSet.has(item.question_id));
  }
  const records = [];

  for (const question of questions) {
    const prompt = buildPrompt(lesson, knowledgePoints, question);
    try {
      const result = await callDoubaoWithRetry({ apiKey, model, prompt, retries });
      await fs.writeFile(
        path.join(rawDir, `${question.question_id}.response.json`),
        JSON.stringify(result.raw, null, 2),
        "utf8",
      );
      const placement = buildStructuredPlacement(result.parsed, lesson, question, pointMap);
      records.push({
        question_id: question.question_id,
        checkpoint: question.checkpoint || "",
        component_label: question.component_label || "",
        status: "ok",
        input_mode: "text_only",
        placement,
      });
    } catch (error) {
      records.push({
        question_id: question.question_id,
        checkpoint: question.checkpoint || "",
        component_label: question.component_label || "",
        status: "failed",
        input_mode: "text_only",
        error: String(error?.message || error),
      });
    }
    if (sleepMs > 0) {
      await new Promise((resolve) => setTimeout(resolve, sleepMs));
    }
  }

  const resultBundle = {
    lesson_id: lessonId,
    lesson_title: lesson.lesson_title,
    model,
    input_mode: "text_only",
    question_count: records.length,
    ok_count: records.filter((item) => item.status === "ok").length,
    failed_count: records.filter((item) => item.status !== "ok").length,
    records,
  };

  await fs.writeFile(path.join(outDir, "placement_results.json"), JSON.stringify(resultBundle, null, 2), "utf8");
  const compact = records
    .filter((item) => item.status === "ok")
    .map((item) => buildCompactRecord({ question_id: item.question_id, checkpoint: item.checkpoint }, item.placement));
  await fs.writeFile(path.join(outDir, "placement_compact.json"), JSON.stringify(compact, null, 2), "utf8");
  await writeSummary(path.join(outDir, "placement_summary.md"), lesson, knowledgePoints, records.filter((item) => item.status === "ok"));

  console.log(
    JSON.stringify(
      {
        out_dir: outDir,
        question_count: resultBundle.question_count,
        ok_count: resultBundle.ok_count,
        failed_count: resultBundle.failed_count,
      },
      null,
      2,
    ),
  );
}

main().catch((error) => {
  console.error(String(error?.stack || error));
  process.exitCode = 1;
});

File: tools/doubao_visual_lesson_placement.py
# Purpose:
# - Python companion for Doubao-based visual lesson placement and markdown summary export.
# - This version is useful when OCR/image preprocessing is easier to express in Python utilities.

from __future__ import annotations

import argparse
import base64
import json
import os
import re
import time
import urllib.error
import urllib.request
from collections import Counter
from pathlib import Path


API_URL = "https://ark.cn-beijing.volces.com/api/v3/chat/completions"
DEFAULT_MODEL = "doubao-seed-2-0-pro-260215"


def read_json(path: Path) -> dict | list:
    return json.loads(path.read_text(encoding="utf-8"))


def ensure_dir(path: Path) -> None:
    path.mkdir(parents=True, exist_ok=True)


def normalize_text(text: str) -> str:
    return re.sub(r"\s+", " ", str(text or "")).strip()


def image_to_data_url(path: Path) -> str:
    suffix = path.suffix.lower()
    mime = {
        ".png": "image/png",
        ".jpg": "image/jpeg",
        ".jpeg": "image/jpeg",
        ".webp": "image/webp",
    }.get(suffix, "image/png")
    b64 = base64.b64encode(path.read_bytes()).decode("ascii")
    return f"data:{mime};base64,{b64}"


def extract_json_block(text: str) -> dict:
    text = str(text or "").strip()
    if not text:
      raise ValueError("empty_model_response")

    if text.startswith("```"):
        text = re.sub(r"^```(?:json)?\s*", "", text)
        text = re.sub(r"\s*```$", "", text)

    try:
        return json.loads(text)
    except json.JSONDecodeError:
        pass

    start = text.find("{")
    if start < 0:
        raise ValueError("json_object_not_found")

    depth = 0
    in_string = False
    escape = False
    for idx in range(start, len(text)):
        ch = text[idx]
        if in_string:
            if escape:
                escape = False
            elif ch == "\\":
                escape = True
            elif ch == '"':
                in_string = False
            continue
        if ch == '"':
            in_string = True
        elif ch == "{":
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth == 0:
                return json.loads(text[start : idx + 1])
    raise ValueError("json_object_not_closed")


def build_candidate_packet(knowledge_points: list[dict]) -> list[dict]:
    packet = []
    for point in knowledge_points:
        packet.append(
            {
                "knowledge_id": point["knowledge_id"],
                "module": point["level_2_module"],
                "min_knowledge_point": point["level_3_min_knowledge_point"],
                "lesson_title": point["lesson_title"],
                "stage": point["stage"],
                "grade": point["grade"],
                "season": point["season"],
            }
        )
    return packet


def build_prompt(lesson: dict, knowledge_points: list[dict], question: dict) -> str:
    candidates = build_candidate_packet(knowledge_points)
    return (
        "你是领世培优一对一数学教研落位助手。"
        "这次不是泛化分类，而是把一道题落到当前既有暑假讲义课次里的最小知识点。"
        "请先看题图，OCR 文字只作为辅助。"
        "不要偷懒复述已有考点名，必须根据题目内容判断。"
        "\n\n"
        "已知外层范围已经锁定：\n"
        f"- 学科：数学\n- 学段：{lesson['stage']}\n- 年级：{lesson['grade']}\n- 季节：{lesson['season']}\n"
        f"- 课次：{lesson['lesson_title']}（{lesson['lesson_id']}）\n"
        "\n候选最小知识点如下，只能从这里面选 Top3：\n"
        f"{json.dumps(candidates, ensure_ascii=False, indent=2)}\n"
        "\n当前题目信息：\n"
        f"- question_id: {question['question_id']}\n"
        f"- 讲义内标签: {question.get('checkpoint', '')}\n"
        f"- 组件: {question.get('component_label', '')}\n"
        f"- 本地题号: {question.get('local_number', '')}\n"
        f"- 页码: {','.join(map(str, question.get('visual_pages', [])))}\n"
        f"- OCR辅助文本: {normalize_text(question.get('text_preview', ''))}\n"
        "\n输出要求：\n"
        "1. 只输出一个 JSON 对象，不要加代码块。\n"
        "2. stage / grade / lesson 三层都要给出判断理由，即使范围已锁定，也要写你为什么确认。\n"
        "3. fine_point.top_candidates 与 final_top3 的 knowledge_id 必须来自候选表。\n"
        "4. confidence 只能写 high / medium / low。\n"
        "5. review_status 只能写 accepted_candidate 或 needs_teacher_review。\n"
        "6. question_reading 要像老师一样概括题型、条件、所求和是否依赖图像。\n"
        "\n严格按下面结构输出：\n"
        "{\n"
        '  "question_id": "...",\n'
        '  "question_reading": {\n'
        '    "core_objects": ["..."],\n'
        '    "conditions": ["..."],\n'
        '    "asked_result": "...",\n'
        '    "visual_dependency": "none|formula|diagram|unknown"\n'
        "  },\n"
        '  "layered_trace": {\n'
        '    "stage": {"prediction": "初中|高中", "confidence": "high|medium|low", "reason": "..."},\n'
        '    "grade": {"top_candidates": [{"grade": "...", "confidence": "...", "reason": "..."}]},\n'
        '    "lesson": {"top_candidates": [{"lesson_id": "...", "lesson_title": "...", "confidence": "...", "reason": "..."}]},\n'
        '    "module": {"top_candidates": [{"module": "...", "confidence": "...", "reason": "..."}]},\n'
        '    "fine_point": {"top_candidates": [{"knowledge_id": "...", "module": "...", "min_knowledge_point": "...", "confidence": "...", "reason": "..."}]}\n'
        "  },\n"
        '  "final_top3": [\n'
        '    {"rank": 1, "knowledge_id": "...", "grade": "...", "lesson_title": "...", "module": "...", "min_knowledge_point": "...", "confidence": "high|medium|low", "teacher_review_note": "..."}\n'
        "  ],\n"
        '  "review_status": "accepted_candidate|needs_teacher_review"\n'
        "}"
    )


def call_model(api_key: str, model: str, prompt: str, image_path: Path, include_image: bool) -> dict:
    user_content: str | list[dict]
    if include_image:
        user_content = [
            {"type": "text", "text": prompt},
            {"type": "image_url", "image_url": {"url": image_to_data_url(image_path)}},
        ]
    else:
        user_content = prompt
    body = {
        "model": model,
        "temperature": 0.1,
        "messages": [
            {
                "role": "system",
                "content": "你是严谨的K12数学教研落位助手。输出必须是JSON对象。视觉优先，OCR只做辅助。",
            },
            {
                "role": "user",
                "content": user_content,
            },
        ],
    }
    request = urllib.request.Request(
        API_URL,
        data=json.dumps(body, ensure_ascii=False).encode("utf-8"),
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(request, timeout=180) as response:
            raw = response.read().decode("utf-8")
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"http_{exc.code}: {detail}") from exc
    except urllib.error.URLError as exc:
        raise RuntimeError(f"network_error: {exc}") from exc

    payload = json.loads(raw)
    content = payload["choices"][0]["message"]["content"]
    parsed = extract_json_block(content)
    return {
        "raw_response": payload,
        "parsed": parsed,
    }


def build_markdown_summary(
    lesson: dict,
    knowledge_points: list[dict],
    results: list[dict],
    out_path: Path,
) -> None:
    top1_counter = Counter()
    needs_review = []
    for item in results:
        parsed = item.get("placement", {})
        top3 = parsed.get("final_top3") or []
        if top3:
            top1_counter[top3[0].get("knowledge_id", "unmatched")] += 1
        if parsed.get("review_status") != "accepted_candidate":
            needs_review.append(item)

    point_map = {point["knowledge_id"]: point for point in knowledge_points}
    lines = [
        f"# 豆包视觉落位结果\n\n",
        f"- 课次：{lesson['lesson_title']}（{lesson['lesson_id']}）\n",
        f"- 学段/年级：{lesson['stage']} / {lesson['grade']}\n",
        f"- 处理题数：{len(results)}\n",
        f"- 需要教师复核：{len(needs_review)}\n\n",
        "## Top1 分布\n\n",
    ]
    for knowledge_id, count in top1_counter.most_common():
        point = point_map.get(knowledge_id, {})
        lines.append(
            f"- {knowledge_id} | {point.get('level_2_module', '')} / {point.get('level_3_min_knowledge_point', '')} ：{count}题\n"
        )
    lines.append("\n## 需要复核\n\n")
    if not needs_review:
        lines.append("- 本轮未出现强制复核题。\n")
    else:
        for item in needs_review:
            parsed = item.get("placement", {})
            top1 = (parsed.get("final_top3") or [{}])[0]
            lines.append(
                f"- {item['question_id']} | {item['checkpoint']} | Top1={top1.get('knowledge_id', '')} | {top1.get('teacher_review_note', '')}\n"
            )
    out_path.write_text("".join(lines), encoding="utf-8")


def main() -> None:
    parser = argparse.ArgumentParser(description="Use Doubao vision to place split lesson questions into lesson-level knowledge points.")
    parser.add_argument("--split-json", required=True)
    parser.add_argument("--knowledge-json", required=True)
    parser.add_argument("--lessons-json", required=True)
    parser.add_argument("--lesson-id", required=True)
    parser.add_argument("--out-dir", required=True)
    parser.add_argument("--model", default=DEFAULT_MODEL)
    parser.add_argument("--api-key", default=os.environ.get("ARK_API_KEY", ""))
    parser.add_argument("--limit", type=int, default=0)
    parser.add_argument("--sleep-seconds", type=float, default=0.3)
    parser.add_argument("--disable-image", action="store_true")
    args = parser.parse_args()

    if not args.api_key:
        raise SystemExit("missing_api_key")

    split_json = Path(args.split_json)
    knowledge_json = Path(args.knowledge_json)
    lessons_json = Path(args.lessons_json)
    out_dir = Path(args.out_dir)
    ensure_dir(out_dir)

    split_data = read_json(split_json)
    all_points = read_json(knowledge_json)
    all_lessons = read_json(lessons_json)
    lesson = next((item for item in all_lessons if item["lesson_id"] == args.lesson_id), None)
    if not lesson:
        raise SystemExit(f"lesson_not_found: {args.lesson_id}")

    knowledge_points = [item for item in all_points if item["lesson_id"] == args.lesson_id]
    if not knowledge_points:
        raise SystemExit(f"knowledge_points_not_found: {args.lesson_id}")

    questions = split_data.get("questions", [])
    if args.limit and args.limit > 0:
        questions = questions[: args.limit]

    input_mode = "text_only" if args.disable_image else "text_plus_image"
    records = []
    raw_dir = out_dir / "raw"
    ensure_dir(raw_dir)

    for idx, question in enumerate(questions, start=1):
        crop_path = Path(question["crop_path"])
        if not crop_path.exists():
            records.append(
                {
                    "question_id": question["question_id"],
                    "checkpoint": question.get("checkpoint", ""),
                    "status": "failed",
                    "error": f"missing_crop: {crop_path}",
                }
            )
            continue

        prompt = build_prompt(lesson, knowledge_points, question)
        try:
            result = call_model(args.api_key, args.model, prompt, crop_path, include_image=not args.disable_image)
            raw_path = raw_dir / f"{question['question_id']}.response.json"
            raw_path.write_text(json.dumps(result["raw_response"], ensure_ascii=False, indent=2), encoding="utf-8")
            records.append(
                {
                    "question_id": question["question_id"],
                    "checkpoint": question.get("checkpoint", ""),
                    "component_label": question.get("component_label", ""),
                    "crop_path": str(crop_path),
                    "input_mode": input_mode,
                    "placement": result["parsed"],
                    "status": "ok",
                }
            )
        except Exception as exc:  # noqa: BLE001
            records.append(
                {
                    "question_id": question["question_id"],
                    "checkpoint": question.get("checkpoint", ""),
                    "component_label": question.get("component_label", ""),
                    "crop_path": str(crop_path),
                    "input_mode": input_mode,
                    "status": "failed",
                    "error": str(exc),
                }
            )
        time.sleep(max(args.sleep_seconds, 0.0))

    summary = {
        "lesson_id": args.lesson_id,
        "lesson_title": lesson["lesson_title"],
        "question_count": len(records),
        "ok_count": sum(1 for item in records if item["status"] == "ok"),
        "failed_count": sum(1 for item in records if item["status"] != "ok"),
        "model": args.model,
        "input_mode": input_mode,
        "records": records,
    }

    json_path = out_dir / "placement_results.json"
    json_path.write_text(json.dumps(summary, ensure_ascii=False, indent=2), encoding="utf-8")
    compact = []
    for item in records:
        parsed = item.get("placement", {})
        top1 = (parsed.get("final_top3") or [{}])[0]
        compact.append(
            {
                "question_id": item["question_id"],
                "checkpoint": item.get("checkpoint", ""),
                "status": item["status"],
                "top1_knowledge_id": top1.get("knowledge_id", ""),
                "top1_module": top1.get("module", ""),
                "top1_min_knowledge_point": top1.get("min_knowledge_point", ""),
                "review_status": parsed.get("review_status", ""),
                "top1_confidence": top1.get("confidence", ""),
            }
        )
    (out_dir / "placement_compact.json").write_text(json.dumps(compact, ensure_ascii=False, indent=2), encoding="utf-8")
    build_markdown_summary(lesson, knowledge_points, [item for item in records if item["status"] == "ok"], out_dir / "placement_summary.md")
    print(json.dumps({"out_dir": str(out_dir), "question_count": len(records), "ok_count": summary["ok_count"], "failed_count": summary["failed_count"]}, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()
